src/IR.hs

```diff
@@ -1,43 +1,96 @@
+{- | IR stage of the compiler pipeline.
+
+The IR stage is organized into 4 distinct substages: 'lower', 'ann2Class',
+'class2Poly', and 'poly2Flat', delineated by each type system used in the
+compiler.
+-}
 module IR where
 
-import qualified Common.Compiler               as Compiler
+import           Common.Compiler
+import           Common.Default                 ( Default(..) )
 
 import qualified Front.Ast                     as A
 
 import qualified IR.IR                         as I
-
 import qualified IR.Types.Annotated            as Ann
-import qualified IR.Types.Classes              as Classes
+import qualified IR.Types.Classes              as Cls
 import qualified IR.Types.Flat                 as Flat
 import qualified IR.Types.Poly                 as Poly
 
 import           IR.ClassInstantiation          ( instProgram )
+import           IR.LambdaLift                  ( liftProgramLambdas )
 import           IR.LowerAst                    ( lowerProgram )
 import           IR.Monomorphize                ( monoProgram )
 import           IR.TypeInference               ( inferProgram )
-import           IR.LambdaLift                  ( liftProgramLambdas )
 
-lowerAst :: A.Program -> Compiler.Pass (I.Program Ann.Type)
-lowerAst = lowerProgram
+import           Control.Monad                  ( when )
+import           System.Console.GetOpt          ( ArgDescr(..)
+                                                , OptDescr(..)
+                                                )
+
+-- | Operation modes for the IR compiler stage.
+data Mode
+  = DumpIR
+  | DumpIRTyped
+  | DumpIRLifted
+  | DumpIRFinal
+  | Continue
+  deriving (Eq, Show)
+
+-- | Compiler options for the IR compiler stage.
+newtype Options = Options { mode :: Mode } deriving (Eq, Show)
+
+instance Default Options where
+  def = Options { mode = Continue }
 
-inferTypes :: I.Program Ann.Type -> Compiler.Pass (I.Program Classes.Type)
-inferTypes = inferProgram
+-- | CLI options for the IR compiler stage.
+options :: [OptDescr (Options -> Options)]
+options =
+  [ Option ""
+           ["dump-ir"]
+           (NoArg $ setMode DumpIR)
+           "Print the IR immediately after lowering"
+  , Option ""
+           ["dump-ir-typed"]
+           (NoArg $ setMode DumpIRTyped)
+           "Print the fully-typed IR after type inference"
+  , Option ""
```

Comment:
This is meaning an option for `DumpIRLifted`?

Reply:
Oops yeah, I'll add that in.